### PR TITLE
Adds functions to operate on StaticArray

### PR DESCRIPTION
**Pull request by TimDeve**
This PR tries to get StaticArray to parity with Array function wise where possible. Functions that creates a new array were skipped. Also adds test where they were missing in `test/array.carp`.

Adds these functions:
- `Array.from-static`
- `StaticArray.reverse!`

Copies the following functions from Array to ArrayStatic:
- `empty?`
- `any?`
- `all?`
- `find`
- `find-index`
- `first`
- `unsafe-first`
- `last`
- `unsafe-last`
- `minimum`
- `maximum`
- `sum`
- `index-of`
- `element-count`
- `predicate-count`
- `aupdate!`
- `swap!`
- `nth`
- `contains?`

### CODE REVIEW

File: core/Array.carp [hunk 1]
```diff
@@ -323,9 +323,9 @@ The trailing elements of the longer array will be discarded.")
 
 If the `index` is out of bounds, return `Maybe.Nothing`")
   (defn nth [xs index]
-  (if (and (>= index 0) (< index (length xs)))
-    (Maybe.Just @(unsafe-nth xs index)) ; the copy will go away with lifetimes
-    (Maybe.Nothing)))
+   (if (and (>= index 0) (< index (length xs)))
```
Comment (hellerve) [inline core/Array.carp:326]: While we’re at it, I think we could even indent by one more space here.

File: core/Array.carp [hunk 2]
```diff
@@ -390,4 +390,11 @@ It will create a copy. If you want to avoid that, consider using [`endo-filter`]
             (set! a (push-back a (Array.slice arr x y)))
             (set! x y)))
         a))
-)
+
+  (doc from-static "Turns a StaticArray into an Array. Copies elements.")
```
Comment (hellerve) [inline core/Array.carp:394]: If you want you can use Markdown syntax here for the types, like this:

```
(doc from-static "Turns a `StaticArray` into an `Array`. Copies elements.")
```

File: core/StaticArray.carp [hunk 3]
```diff
@@ -1,5 +1,23 @@
 (defmodule StaticArray
 
+  (doc map! "Maps a function over the static array `xs`, mutating it in place. The difference to Array.endo-map (which does the same thing internally) is that this function takes a ref (since you can never have static arrays as values) and that it returns ().")
+  (defn map! [xs f]
```
Comment (hellerve) [inline core/StaticArray.carp:4]: For convenience, could we return `xs` here anyway?

Comment (eriksvedang) [reply core/StaticArray.carp:4]: I don't think that's very convenient since you'll have to `(ignore ...)` that often, no?

Comment (TimDeve) [reply core/StaticArray.carp:4]: Don't believe so as every static array literals are (Ref StaticArray)

Comment (hellerve) [reply core/StaticArray.carp:4]: Hm, okay! I just thought that a `map!` that doesn’t return anything looks a little weird, but I guess that makes sense.

Comment (eriksvedang) [reply core/StaticArray.carp:4]: Since all other `!` function returns `()` I think this one should too.

Comment (hellerve) [reply core/StaticArray.carp:4]: Sure, this can be marked as resolved from my end :)

**Comment by TimDeve** [comment]
Made the requested edits, sorry about the random indentation changes, that would be Parinfer acting up.

File: core/StaticArray.carp [hunk 4]
```diff
@@ -1,5 +1,23 @@
...
+             j (Int.dec (length a))]
+      (while (Int.< i j)
+        (let-do [tmp @(unsafe-nth a i)]
+          (aset! a i @(unsafe-nth a j))
+          (set! i (Int.inc i))
+          (aset! a j tmp)
+          (set! j (Int.dec j))))))
+
+  ;; NOTE: Following function declarations are copy of the ones in Array
+  ;; could we share inmplementation between Array and StaticArray?
```
Comment (eriksvedang) [inline core/StaticArray.carp:20]: typo "inmplementation"

Comment (eriksvedang) [reply core/StaticArray.carp:20]: I guess that whole second line can be removed since we've kinda resolved that in chat, for now?

Comment (TimDeve) [reply core/StaticArray.carp:20]: Ok, will do.

**Comment by eriksvedang** [comment]
The address sanitizer finds an error with the first test case (for `map!`), it's a free after use error since the ref `arr` is returned and the static array lives inside the `let`. This is actually detected when I paste the code into a separate function (only on the most recent commit on `master` though, this was not detected properly until today). 

For some reason it's still not detected in the test case though, so I need to fix that bug with the reference checker.

Please go through the test cases an look if this error pops up in other places. You might have to rearrange the code a bit for it to work, I'm afraid. Also, do you have the Clang address sanitizer? It will help you find these errors locally.

**Comment by TimDeve** [comment]
For some reason I thought that it was fair game to return from a let but I guess it's a block {} in c so the automatic memory gets freed.

**Comment by eriksvedang** [comment]
Yeah, sorry about the compiler not catching this (I'm on it!).

**Comment by TimDeve** [comment]
Fixed the sanitiser errors

**Comment by eriksvedang** [comment]
Hoorray! Thanks for your hard work, this will be super useful. Looking forward to seeing what you can do with it on the raspi!

**Pull request merged**